src/object/cli_obj.c

```diff
@@ -496,92 +496,131 @@ obj_shard_tgts_query(struct dc_object *obj, uint32_t map_ver, uint32_t shard,
 	int				 rc;
 
-	D_ASSERT(*shard_cnt >= 1);
-	if (!srv_io_dispatch || *shard_cnt == 1)
-		return 0;
-
-	/* select one leader shard and generate the fw_shard_tgts array */
-	rc = obj_grp_leader_get(obj, *start_shard, map_ver);
-	if (rc < 0) {
-		D_ERROR(DF_OID" no valid shard, rc %d.\n",
-			DP_OID(obj->cob_md.omd_id), rc);
-		return rc;
-	}
-	lead_shard = rc;
-	if (tgt_set) {
-		for (i = 0; i < *shard_cnt; i++)
-			if (*start_shard + i != lead_shard &&
-			    tgt_set & (1UL << i))
-				new_fw_cnt++;
+	D_ASSERT(shard_cnt >= 1);
+	grp_size = shard_cnt / grp_nr;
+	D_ASSERT(grp_size * grp_nr == shard_cnt);
+	req_tgts->ort_srv_disp = srv_io_dispatch && grp_size > 1;
+
+	if (tgt_set != 0) {
+		D_ASSERT(grp_nr == 1);
+		shard_nr = 0;
+		for (i = 0; i < shard_cnt; i++)
+			if (tgt_set & (1UL << i))
+				shard_nr++;
 	} else {
-		new_fw_cnt = *shard_cnt - 1;
-	}
-	if (shard_tgts != NULL && *fw_cnt != new_fw_cnt) {
-		/* fw_cnt possibly changed per progressive layout */
-		D_FREE(shard_tgts);
-		*fw_shard_tgts = NULL;
-		shard_tgts = NULL;
-		*fw_cnt = 0;
-	}
-	if (shard_tgts == NULL) {
-		D_ALLOC_ARRAY(shard_tgts, new_fw_cnt);
-		if (shard_tgts == NULL)
-			return -DER_NOMEM;
-		*fw_shard_tgts = shard_tgts;
-		*fw_cnt = new_fw_cnt;
-	}
-	for (i = 0, j = 0; i < *shard_cnt; i++) {
-		uint32_t	shard;
-
-		shard = *start_shard + i;
-		if (shard == lead_shard || (tgt_set && !(tgt_set & 1UL << i)))
-			continue;
+		shard_nr = shard_cnt;
+	}
 
-		rc = obj_shard_tgts_query(obj, map_ver, shard,
-					  shard_tgts + j);
-		j++;
-		if (rc != 0) {
-			D_ERROR(DF_OID" obj_shard_tgts_query failed, "
-				"rc %d.\n", DP_OID(obj->cob_md.omd_id),
-				rc);
-			return rc;
+	if (shard_nr > OBJ_TGT_INLINE_NR) {
+		if (req_tgts->ort_shard_tgts != NULL &&
+		    req_tgts->ort_grp_nr * req_tgts->ort_grp_size != shard_nr) {
+			/* shard_nr possibly changed per progressive layout */
+			if (req_tgts->ort_shard_tgts !=
+				req_tgts->ort_tgts_inline)
+				D_FREE(req_tgts->ort_shard_tgts);
+			req_tgts->ort_shard_tgts = NULL;
 		}
+		if (req_tgts->ort_shard_tgts == NULL) {
+			D_ALLOC_ARRAY(req_tgts->ort_shard_tgts, shard_nr);
+			if (req_tgts->ort_shard_tgts == NULL)
+				return -DER_NOMEM;
+		}
+	} else {
+		if (req_tgts->ort_shard_tgts != NULL &&
+		    req_tgts->ort_shard_tgts != req_tgts->ort_tgts_inline)
+			D_FREE(req_tgts->ort_shard_tgts);
+		req_tgts->ort_shard_tgts = req_tgts->ort_tgts_inline;
 	}
+	req_tgts->ort_grp_nr = grp_nr;
+	req_tgts->ort_grp_size = grp_size;
 
-	*start_shard = lead_shard;
-	*shard_cnt = 1;
+	for (i = 0; i < grp_nr; i++) {
+		shard_idx = start_shard + i * grp_size;
+		tgt = req_tgts->ort_shard_tgts + i * grp_size;
+		if (req_tgts->ort_srv_disp) {
+			rc = obj_grp_leader_get(obj, shard_idx, map_ver);
+			if (rc < 0) {
+				D_ERROR(DF_OID" no valid shard, rc %d.\n",
+					DP_OID(obj->cob_md.omd_id), rc);
+				return rc;
+			}
+			leader_shard = rc;
+			rc = obj_shard_tgts_query(obj, map_ver, leader_shard,
+						  tgt++);
```

Comment:
just a question, so the first target of each group is always to leader right?

Reply:
right, there are some comments at struct obj_req_tgts' definition.